feat(har): add more size and http metadata

mxschmitt opened the pull request:
Changes:

- adds `request.bodySize` (same impl than [in Chromium](https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/models/har/Log.ts;l=378-388?q=f:front_end%20-f:out%2F%20responseHeadersText)
- adds `response.bodySize`
- adds `response.headersSize`
- adds `response._transferSize`

Supersedes #6695

yury-s commented on src/server/supplements/har/har.ts:
```diff
@@ -81,6 +81,7 @@ export type Response = {
   redirectURL: string;
   headersSize: number;
   bodySize: number;
+  _transferSize: number;
```
why not name it `transferSize` right away? Is it because the field is non-standard?

mxschmitt replied:
Exactly, its a chromium only field and not yet part of the official spec.

yury-s commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -20,6 +20,9 @@ import { helper } from '../../helper';
 import * as har from './har';
+import * as types from '../../types';
+
+const FALLBACK_HTTP_VERSION = 'http/1.1';
```
I believe HTTP (capital case) is more common and that is also what it was before, let's keep it?

yury-s commented on src/server/network.ts:
```diff
@@ -311,8 +311,10 @@ export class Response extends SdkObject {
   private _serverAddrPromiseCallback: (arg?: RemoteAddr) => void = () => {};
   private _securityDetailsPromise: Promise<SecurityDetails|undefined>;
   private _securityDetailsPromiseCallback: (arg?: SecurityDetails) => void = () => {};
+  _protocol: string | undefined;
```
_httpVersion ?

yury-s commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -168,29 +173,65 @@ export class HarTracer {
+
+    if (!this._options.omitContent) {
+      const promise = response.body().then(buffer => {
+        harEntry.response.content.text = buffer.toString('base64');
```
nit: assign `harEntry.response.content`  to a variable for better readability?

yury-s commented on src/server/webkit/wkPage.ts:
```diff
@@ -1043,7 +1043,11 @@ export class WKPage implements PageDelegate {
       });
-      response._requestFinished(helper.secondsToRoundishMillis(event.timestamp - request._timestamp));
+      const { responseBodyBytesReceived, responseHeaderBytesReceived } = event.metrics || {};
+      const transferSize = responseBodyBytesReceived && responseHeaderBytesReceived ? responseBodyBytesReceived + responseHeaderBytesReceived : undefined;
```
It doesn't seem to account for compression, is it intentional?

mxschmitt replied:
Yes afaik, but let me verify it later against DevTools.

yury-s replied:
I believe it should be equal to responseHeaderBytesReceived if the response has empty body.

yury-s commented on tests/har.spec.ts:
```diff
@@ -260,6 +276,47 @@ it('should calculate time', async ({ contextFactory, server }, testInfo) => {
+  expect(log.entries[1].response._transferSize).toBe(6323);
+});
+
+it('should have -1 transportSize when its a failed request', async ({ contextFactory, server }, testInfo) => {
```
transportSize => _transferSize (here and above)?

yury-s commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -128,27 +133,28 @@ export class HarTracer {
         content: {
           size: -1,
-          mimeType: request.headerValue('content-type') || 'application/octet-stream',
+          mimeType: request.headerValue('content-type') || 'x-unknown',
```
is 'x-unknown' a thing? I'd rather use an empty string to indicate it's unknown or keep 'octet-stream' as it's one of the least specific values

mxschmitt replied:
Thats how DevTools is doing it: https://source.chromium.org/chromium/chromium/src/+/main:third_party/devtools-frontend/src/front_end/models/har/Log.ts;l=237?q=f:front_end%20-f:out%2F%20responseHeadersText

yury-s commented on src/server/supplements/har/harTracer.ts:
```diff
@@ -168,29 +174,67 @@ export class HarTracer {
       content: {
         size: -1,
-        mimeType: response.headerValue('content-type') || 'application/octet-stream',
+        mimeType: response.headerValue('content-type') || 'x-unknown',
```
ditto

mxschmitt committed:
feat(har): add more size and http metadata

mxschmitt committed:
skip transport size tests on ubuntu WK

tnolet committed:
feat(har): records sizes separate from payload base64-ified content

Recording the size metadata is useful in all cases and should not be linked to recording the base64
version of response bodies.

mxschmitt committed:
nits
